pkg/edit/editor_test.go

```diff
@@ -27,14 +27,102 @@ func TestEditor_DoesNotAddEmptyCommandToHistory(t *testing.T) {
 	testCommands(t, f.Store /* no commands */)
 }
 
-func TestEditor_DoesCommandWithLeadingSpaceToHistory(t *testing.T) {
-	f := setup()
-	defer f.Cleanup()
+func TestEditor_TestBeforeAddCmd(t *testing.T) {
+	cases := []struct {
+		name        string
+		rc          string
+		input       string
+		wantHistory []string
+	}{
+		// TODO: Enable the following two tests once error output can
+		// be tested.
+		// {
+		// 	name:        "non callable item",
+		// 	rc:          "edit:add-cmd-filters = [$false]",
+		// 	input:       "echo\n",
+		// 	wantHistory: []string{"echo"},
+		// },
+		// {
+		// 	name:        "callback outputs nothing",
+		// 	rc:          "edit:add-cmd-filters = [[_]{}]",
+		// 	input:       "echo\n",
+		// 	wantHistory: []string{"echo"},
+		// },
+		{
+			name:        "callback outputs true",
+			rc:          "edit:add-cmd-filters = [[_]{ put $true }]",
+			input:       "echo\n",
+			wantHistory: []string{"echo"},
+		},
+		{
+			name:        "callback outputs false",
+			rc:          "edit:add-cmd-filters = [[_]{ put $false }]",
+			input:       "echo\n",
+			wantHistory: nil,
+		},
+		{
+			name:        "false-true chain",
+			rc:          "edit:add-cmd-filters = [[_]{ put $false } [_]{ put $true }]",
+			input:       "echo\n",
+			wantHistory: nil,
+		},
+		{
+			name:        "true-false chain",
+			rc:          "edit:add-cmd-filters = [[_]{ put $true } [_]{ put $false }]",
+			input:       "echo\n",
+			wantHistory: nil,
+		},
+		{
+			name:        "positive",
+			rc:          "edit:add-cmd-filters = [[cmd]{ ==s $cmd echo }]",
+			input:       "echo\n",
+			wantHistory: []string{"echo"},
+		},
+		{
+			name:        "negative",
+			rc:          "edit:add-cmd-filters = [[cmd]{ ==s $cmd echo }]",
+			input:       "echo x\n",
+			wantHistory: nil,
+		},
+		{
+			name:        "default value",
+			rc:          "",
+			input:       " echo\n",
+			wantHistory: nil,
+		},
+	}
 
-	feedInput(f.TTYCtrl, " echo\n")
-	f.Wait()
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			f := setup(rc(c.rc))
+			defer f.Cleanup()
 
-	testCommands(t, f.Store /* no commands */)
+			feedInput(f.TTYCtrl, c.input)
+			f.Wait()
+
+			testCommands(t, f.Store, c.wantHistory...)
+		})
+	}
+
+	t.Run("short-circuiting behavior", func(t *testing.T) {
+		c := `
+edit:add-cmd-filters = [
+    [_]{ put $false }
+    [_]{ edit:add-cmd-filters = []; put $true }
```

Comment:
I'd rather you test the side effect in a more straightforward way, for example by setting a variable, and test the value of the variable using the `testGlobal` helper.